### PR TITLE
add PEFT adapter support for mcore gpt path

### PR DESCRIPTION
**Pull request by blahBlahhhJ**
Add PEFT adapter support for mcore gpt path.

contains a nemo.core change to accept kwargs to pass into hydra instantiate.

currently only support Ptuning and Lora now.

tested on llama2 7b TP SP PP DP

### CODE REVIEW

File: nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    from megatron.core.transformer.custom_layers.transformer_engine import TELinear
+
+    HAVE_MEGATRON_CORE = True
+
+except (ImportError, ModuleNotFoundError):
+
+    HAVE_MEGATRON_CORE = False
+
+    logging.info("Megatron-core is required to use McoreAdapterModuleMixin.")
+    raise RuntimeError("McoreAdapterModuleMixin can not run without Megatron-core.")
```
Comment (blahBlahhhJ) [inline nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py:24]: didn't find a good place to raise this error. Should I do it here? @ericharper

File: nemo/collections/nlp/models/language_modeling/megatron_gpt_peft_models.py [hunk 2]
```diff
@@ -55,16 +60,34 @@ def init_peft_modules(self):
...
+            # if there's an infinite loop here, check that the added adapter modules doesn't use the same key as mcore targets.
+            if self.mcore_gpt:
                 for peft_key in self.peft_name_keys:
-                    peft_cfg = self.name_key_to_cfg[peft_key]
-                    if model_utils.import_class_by_path(peft_cfg._target_) in module.get_accepted_adapter_types():
-                        module.add_adapter(
-                            name=peft_key, cfg=peft_cfg,
-                        )
+                    for mcore_target, mcore_mixin in self.name_key_to_mcore_mixins[peft_key]:
+                        if name.split(".")[-1] == mcore_target:
```
Comment (blahBlahhhJ) [inline nemo/collections/nlp/models/language_modeling/megatron_gpt_peft_models.py:73]: should I do a suffix match or a full regex match?

e.g. the template for lora mcore target will be 
```
"model\.decoder\.layers\.[0-9]+\.self_attention\.linear_qkv"
```

**Comment by blahBlahhhJ** [comment]
update: will pull some general PEFT changes out to merge to main first, and then rebase this branch after those were merged.
#7287 
#7285

File: nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    module.__class__ = mcore_mixin
+    module.mcore_register_adapters()
+
+
+class McoreAdapterModuleMixin(adapter_mixins.AdapterModuleMixin):
+    def mcore_register_adapters(self):
+        raise NotImplementedError("Mcore mixins should implement setup_adapters on a subclass of MyBase")
+
+
+class LoraTELinear(TELinear, McoreAdapterModuleMixin):
```
Comment (arendu) [inline nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py:27]: Could we use the [SelfAttention class from Megatron-LM ](https://github.com/NVIDIA/Megatron-LM/blob/f24fac4ed0dcf0522056521a93445d9a82f501a9/megatron/core/transformer/attention.py#L245) and implement the [get_query_key_value_tensors](https://github.com/NVIDIA/Megatron-LM/blob/f24fac4ed0dcf0522056521a93445d9a82f501a9/megatron/core/transformer/attention.py#L266C20-L266C20) here instead of going to the lower level TELinear? It is a few more lines of code but is gives me a more general recipe for implementing future PEFT methods. LoRA is a bit of an outlier because it can be fully encapsulated into a low-level Linear layer.

**Comment by arendu** [comment]
Thanks, @blahBlahhhJ this looks like a great start with some useful key ingredients -- eg. class swapping with NeMo's subclass of Mcore is great!. I left a comment [here](https://github.com/NVIDIA/NeMo/pull/7276/files#r1301979030) which would be useful to address.

File: nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py (hunk 3 above)
Comment (blahBlahhhJ) [reply nemo/collections/nlp/modules/common/megatron/adapters/mcore_mixins.py:27]: I've seen people injecting lora on not only `linear_qkv` but also other linears in the attention, although nemo only applies lora on `linear_qkv` for now...

File: nemo/collections/nlp/modules/common/megatron/adapters/parallel_adapters.py [hunk 4]
```diff
@@ -322,7 +322,7 @@ def __init__(
         gradient_accumulation_fusion = False
         # (@adithyare) the persistent=False will not pollute the indices into the state_dict of this module.
         self.register_buffer("indices", torch.LongTensor(list(range(self.virtual_tokens))), persistent=False)
-        self.embedding = torch.nn.Embedding(self.virtual_tokens, self.embedding_dim)
+        self.prompt_embedding = torch.nn.Embedding(self.virtual_tokens, self.embedding_dim)
```
Comment (arendu) [inline nemo/collections/nlp/modules/common/megatron/adapters/parallel_adapters.py:325]: is this change required? if not, could we leave it as is.
